Remove time-zone dependent call when casting a date via toLocalDate in HiveIcebergFilterFactory

kbendick opened the pull request:
For me, the test `TestHiveIcebergFilterFactory#testDateType` is failing locally, though it doesnt fail in the CI. So whether or not this needs to be changed is debatable, but this test fails for me when I run it locally.

I've changed the test to use `Date.valueOf("2015-11-12")`, which matches the `Literal.of("2015-11-12")` more closely anyway.

I've also made the string literal for the date its own named argument, under the hopes that possibly these tests could be parameterized some day.

When  I test this on my local, the output of the failing test is 
```
expected:<16751> but was:<16750>
Expected :16751
Actual   :16750
...
org.apache.iceberg.mr.hive.TestHiveIcebergFilterFactory > testDateType FAILED
    java.lang.AssertionError: expected:<16751> but was:<16750>
```

kbendick commented:
cc @shardulm94 for your thoughts on this and whether or not you experience this problem on your local machine.

kbendick commented:
I spoke to Shardul offline and got confirmation that this test is failing on their local machine as well.

Admittedly it's not failing on CI, but it's not the developer experience for anybody who runs the build and test themselves vs offloading to Travis.

rdblue commented on mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergFilterFactory.java:
```diff
@@ -207,8 +206,8 @@ public void testBooleanType() {
     SearchArgument.Builder builder = SearchArgumentFactory.newBuilder();
-    Date gmtDate = new Date(LocalDate.of(2015, 11, 12).atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli());
-    SearchArgument arg = builder.startAnd().equals("date", PredicateLeaf.Type.DATE, gmtDate).end().build();
+    Date date = Date.valueOf("2015-11-12");
```
We need to avoid using classes that will change behavior based on time zone to parse the value of a date or timestamp. This method was purposely not used. The problem here is probably that although the instant value is correct, the Date uses the JVM time zone somehow.

kbendick replied:
I had a feeling this method was purposely not used, but I wasn't sure of a better option so I figured I'd open the PR and then let somebody point it out. I also figured that the flakey test was indicative of more than just a flakey test. I'll apply your suggestions this evening after work.

kbendick replied:
I reverted this to the previous code. It's running on my local now with the patch for `daysFromDate`.

rdblue commented:
The problem here is that `HiveIcebergFilterFactory.daysFromDate` calls `date.toLocalDate`, which is zone-dependent. So although we set the underlying date value correctly, it is corrupted by calling `toLocalDate`. We need to fix that method to translate correctly.

kbendick commented:
Cool. I can push a fix using your change after work today 👍 . @rdblue

kbendick commented:
Ok. I reverted the unit test changes and updated the code. So it seems to be passing for me now too.

However, when I run this, the value of `gmtDate` is still `2015-11-11` if I print it right after it's created.

The values now for expected and actual are both 16751. I also printed out the Date value generated in `leafToLiteral` and it is also `2015-11-11`.

 Is this expected due to the use of exclusive bounds or something?

```
gmtDate:2015-11-11
The constant date is 2015-11-11
The date in leafToLiteral using daysFromDate is 16751
expected:ref(name="date") == 16751
actual:ref(name="date") == 16751
```

rdblue commented:
The toString representation from Date is in the local zone. I don't think that it matters what the toString shows, as long as the actual value it gets converted to for Iceberg is correct. That's because Hive is responsible for translating "2020-11-12" into a Date however it chooses, but Iceberg should always use the date of the instant in UTC.

kbendick commented:
The CI failed for Java 8 a transient HMS connection issue similar to this one. I get these intermittently on my local but have never had them on CI.

```
org.apache.iceberg.mr.hive.TestHiveIcebergStorageHandlerWithHiveCatalog > testScanTableORC FAILED
    java.lang.IllegalArgumentException: Failed to executeQuery Hive query SHOW TABLES: Error while compiling statement: FAILED: SemanticException org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)

        Caused by:
        org.apache.hive.service.cli.HiveSQLException: Error while compiling statement: FAILED: SemanticException org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)

            Caused by:
            org.apache.hadoop.hive.ql.parse.SemanticException: org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)

                Caused by:
                org.apache.hadoop.hive.ql.metadata.HiveException: org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)

                    Caused by:
                    org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)

                        Caused by:
                        java.net.SocketException: Broken pipe (Write failed)

    org.apache.hadoop.hive.ql.metadata.HiveException: org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)

        Caused by:
        org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)

            Caused by:
            java.net.SocketException: Broken pipe (Write failed)
```

kbendick commented:
> The toString representation from Date is in the local zone. I don't think that it matters what the toString shows, as long as the actual value it gets converted to for Iceberg is correct. That's because Hive is responsible for translating "2020-11-12" into a Date however it chooses, but Iceberg should always use the date of the instant in UTC.

Good point. Been a while since I've debug printed a Date manually like that.

kbendick commented:
I added a bunch of temporary commits and finally got CI to work, but then reset hard back to the old spot and now it shows as failing again.

I think upping the ulimit when running the tests might possibly help?

kbendick commented:
I think this might be why it's failing: https://github.com/apache/iceberg/pull/1585 

The update to the test memory in that PR is possibly not in my branch.

rdblue commented:
@pvary and @marton-bod, looks like we have some flaky behavior with Hive tests. Here's the failure from @kbendick's comment above:

> ```
> org.apache.iceberg.mr.hive.TestHiveIcebergStorageHandlerWithHiveCatalog > testJoinTablesParquet FAILED
>     java.lang.IllegalArgumentException: Failed to executeQuery Hive query SHOW TABLES: Error while compiling statement: FAILED: SemanticException org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)
>         Caused by:
>         org.apache.hive.service.cli.HiveSQLException: Error while compiling statement: FAILED: SemanticException org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)
>             Caused by:
>             org.apache.hadoop.hive.ql.parse.SemanticException: org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)
>                 Caused by:
>                 org.apache.hadoop.hive.ql.metadata.HiveException: org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)
>                     Caused by:
>                     org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)
>                         Caused by:
>                         java.net.SocketException: Broken pipe (Write failed)
>     org.apache.hadoop.hive.ql.metadata.HiveException: org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)
>         Caused by:
>         org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)
>             Caused by:
>             java.net.SocketException: Broken pipe (Write failed)
> ```

This is probably caused by too many open connections at once. Can you help us double-check the use of client pools in Hive? Are client pools reused across tests with HiveRunner?

rdblue commented:
Since the test failures are unrelated, I'm going to merge this fix. Thanks @kbendick!
